### PR TITLE
Supersig - Milestone 1 (MVP) completed

### PR DESCRIPTION
**Pull request by decentration**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/959

### CODE REVIEW

**Comment by decentration** [comment]
@laboon hi Bill, just checking hat will it take to get this PR reviewed? 

regarding the two failed checks:
-  the author check failed because I seemed to have PR'd the initial front proposal from my more private account @ramseykes. 
- the google sheet update failed, but i am not sure, i tried to submit the google form twice, but still no green tick.

File: deliveries/pallet_supersig_milestone-1_decentration.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Milestone Delivery :mailbox:
+
+> ⚡ Only the GitHub account that submitted the application is allowed to submit milestones. 
+> 
+> Don't remove any of the mandatory parts presented in bold letters or as headlines! Lines starting with `>`, such as this one, can be removed.
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).**  
+
+* **Application Document:** Please, provide a link to the merged contract (the `.md` file in the [applications](https://github.com/w3f/Grants-Program/tree/master/applications) directory).
```
Comment (takahser) [inline deliveries/pallet_supersig_milestone-1_decentration.md:9]:
```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/pallet_supersig.md
```

**Review by takahser — CHANGES_REQUESTED** [review]
@decentration you're supposed to add the link to the application document, that's probably why the actions failed :)

File: deliveries/pallet_supersig_milestone-1_decentration.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+| 0c. | Testing Guide | https://github.com/kabocha-network/pallet_supersig/tree/polkadot-v0.9.28/src/tests | cargo test `INSERT FILENAME` |
+| 0d. | Docker |  https://github.com/decentration/substrate-supersig-template/tree/v0.9.28 | |
+| 0e. | Article and video | https://decentration.medium.com/supersig-like-multisig-but-with-superpowers-86b9ce0412f6, https://www.loom.com/share/dbcaa6319b1a4644aacb709aa0e38783 | |
+| 1. | Substrate module: pallet_supersig | https://github.com/kabocha-network/pallet_supersig/tree/polkadot-v0.9.28/ | |
+| 2. | Benchmarking | The pallet is benchmarked. |https://github.com/kabocha-network/pallet_supersig/blob/polkadot-v0.9.28/src/benchmarking.rs | |
+
+**Additional Information**
+> Any further comments on the milestone that you would like to share with us.
+
+This milestone was completed, we also used half of milestone 3 (Feedback and improvements), to iterate supersig and add RPC modules to make it much more usable. The other half of milestone 3 will be used for milestone 2 (UI development).
```
Comment (takahser) [inline deliveries/pallet_supersig_milestone-1_decentration.md:49]: You're supposed to create 1 delivery document for each delivery. Hence, please update this doc to reflect the changes for only M1.

File: deliveries/pallet_supersig_milestone-1_decentration.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+> Don't remove any of the mandatory parts presented in bold letters or as headlines! Lines starting with `>`, such as this one, can be removed.
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).**  
+
+* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/pallet_supersig.md
+
+**Supersig link**
+https://github.com/w3f/Grants-Program/blob/master/applications/pallet_supersig.md
+
+
```
Comment (takahser) [inline deliveries/pallet_supersig_milestone-1_decentration.md:14]:
```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/pallet_supersig.md
```

File: deliveries/pallet_supersig_milestone-1_decentration.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+* **Milestone Number:** 1, and half of 3.
```
Comment (takahser) [inline deliveries/pallet_supersig_milestone-1_decentration.md:15]:
```suggestion
* **Milestone Number:** 1
```

File: deliveries/pallet_supersig_milestone-1_decentration.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+Secondly, there is also the "master" user feature. A master user always has 50% of the voting power. This means that if a master user is involved in a transaction, only one other person needs to sign in order to approve the transaction for execution. (More than one person can be a master user). 
+
+Supersig is now part of Kabocha parachain's main net on Kusama. 
+
+Milestone 1 is about building pallet_supersig's first version. The pallet now allows you to do the abovementioned core functions. 
+
+We have also apportioned half of milestone 3's feedback and imrprovements, section to: 
+a) create RPC module to make it much easier to query the storage.  
+b) change from Account nonce to AccountId to for usability.
```
Comment (takahser) [inline deliveries/pallet_supersig_milestone-1_decentration.md:29]: Feel free to accommodate this into M3 instead.

**Review by takahser — CHANGES_REQUESTED** [review]
@decentration see my comments bellow. ⬇️
Also, please accept my apologies for the delay, we didn't see the delivery earlier because the actions failed. Somebody from the team will take a look at the application soon.

**Comment by decentration** [comment]
> @decentration see my comments bellow. ⬇️ Also, please accept my apologies for the delay, we didn't see the delivery earlier because the actions failed. Somebody from the team will take a look at the application soon.

Thank you, this was my error, I must have overlooked the process of adding the link to application on the google form. 

I have made the requested change to the milestone. 

Shall I re-submit the invoice to separate out the milestones? And would it be easier to email it to someone?

**Comment by takahser** [comment]
@decentration 

> Thank you, this was my error, I must have overlooked the process of adding the link to application on the google form.

Yes np, now it worked :)

> Shall I re-submit the invoice to separate out the milestones?

Yes, could you please resubmit it, so that it reflects M1 only? Also, could you remove the M3-related content from the M1 delivery, as discussed?

**Comment by decentration** [comment]
hi @takahser Seraya, milestone 1 has been separated from 3, and google form has been [resubmitted](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/formResponse).

**Comment by takahser** [comment]
@decentration ok thanks for updating, looks good now!

File: deliveries/pallet_supersig_milestone-1_decentration.md (hunk 2 above)
Comment (decentration) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: hi @aminubi, i just pushed the updated branch to the grant milestone delivery which is now this branch: https://github.com/decentration/substrate-supersig-template/tree/v0.9.28-new, which compiles and runs.

Comment (aminubi) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: @decentration another Error: Service(Keystore(Io(Os { code: 13, kind: PermissionDenied, message: "Permission denied" }))) when running with docker image.

Comment (decentration) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: @aminubi try this `docker run -p 9944:9944 -p 30333:30333 -p 9933:9933 -p 9615:9615  docker.io/decentration/supersig-node:latest --dev --port 30333 --ws-port 9944 --rpc-port 9933`. I have updated the readme to contain this command. It should start a local node template, and publish to the correct ports so you can interact with it from polkadot js apps.

Comment (decentration) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: though if you are on macOS you may get some problems when connecting via polkadot js apps UI. Trying to get to the bottom of it, but it seems its a wider Docker MacOS issue.

Comment (decentration) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: here is the issue from [docker](https://github.com/docker/for-mac/issues/3487#issuecomment-509176608). But i can confirm it works for Mac Intel chip.

Comment (decentration) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: hey @aminubi, how is the docker running?

Comment (decentration) [reply deliveries/pallet_supersig_milestone-1_decentration.md]: I'll mark this as resolved, as the initial request has been changed, and the docker is also working. If you still have an issue please let me know.

**Comment by takahser** [comment]
@decentration we're going to give you feedback to your delivery soon. 👍

**Review by takahser — CHANGES_REQUESTED** [review]
@decentration I took a closer look at your delivery and have a couple of questions:

- in your grant application the `create_supersig` function included a `threshold: u128` argument that's missing in your [actual delivery](https://github.com/w3f/Grants-Program/blob/4f42b509823a6690144243241b271ed819dffa0b/applications/pallet_supersig.md?plain=1#L152). Why is that?
- I found your docker image on [dockerhub](https://hub.docker.com/r/decentration/supersig-node) however, I couldn't find the source. Also, the [node repo](https://github.com/decentration/substrate-supersig-template) contains the original _Dockerfile_ and _docker-compose.yml_ which was confusing to me, since in your README you reference your own docker image instead.
- Although the node seems to be running fine, I can't connect to it for some reason. Any idea, why?
    image

I added my evaluation [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/supersig_1_takahser.md). Feel free to have a look.

**Comment by decentration** [comment]
sorry didn't mean to remove request for review from @aminubi...

**Comment by decentration** [comment]
hi @takahser Seraya, did you manage to get over the Docker hurdle? 

Also it seems that Aminu is unavailable as he has not responding in over a week.

**Comment by decentration** [comment]
Hi @takahser can I check on your status? its been a few weeks now, and there are relatively sparse responses. I’m hoping that after you get past the docker hurdle then you will find a much smoother reviewal of this milestone. I am here to respond pretty rapidly to any of your questions. Would it be impossible to bump this milestone as a priority?

**Comment by decentration** [comment]
hey @takahser @laboon I think this milestone review is getting lost in the pile... There was a slight docker issue because m1/m2 macs have a known issue, but everything else should be smooth sailing.

**Comment by takahser** [comment]
@decentration sorry I didn't yet find the time to take another look, but I'm planning to give you feedback within the next few days.

**Comment by decentration** [comment]
hey @takahser seraya, how are you getting on? let me know if you have any questions.

**Comment by decentration** [comment]
hey @takahser, i hope youre great! It's been a week since we last spoke, just checking in to ask when you will help to complete this process? I am just wondering how long this needs to take? Seems like it could be done in less than a week. Would be fair perhaps to give this a little more priority now we are in the fifth week.

**Comment by takahser** [comment]
@decentration sorry for the delay, I still didn't find the time to check again. I'll check tomorrow and get back to you!

**Review by takahser — CHANGES_REQUESTED** [review]
@decentration I finally managed to take another look, here are my findings:

- I noticed that the instructions for running the tests are [still missing](https://github.com/w3f/Grant-Milestone-Delivery/blob/a24efce4260285bbf535f728dc6851e0b8194f96/evaluations/supersig_1_takahser.md?plain=1#L72), so feel free to add them.
- Also, I'm [still not able](https://github.com/w3f/Grant-Milestone-Delivery/blob/a24efce4260285bbf535f728dc6851e0b8194f96/evaluations/supersig_1_takahser.md?plain=1#L66) to access your node in polkadot.js apps (same error). As [suggested](https://github.com/decentration/substrate-supersig-template/tree/46c02b54b24a28872238738ec1746a45c72563e4#run-in-docker) in your docs, I started the node using the following command:

  ```bash
  docker run -p 9944:9944 -p 30333:30333 -p 9933:9933 -p 9615:9615 docker.io/decentration/supersig-node:latest --dev --port 30333 --ws-port 9944 --rpc-port 9933 --unsafe-ws-external --rpc-cors all
  ```

Could you check again? Is it working on your machine?
I'm currently not using an M1, instead I'm using an Ubuntu machine.

**Comment by decentration** [comment]
hey @takahser 

> I noticed that the instructions for running the tests are [still missing](https://github.com/w3f/Grant-Milestone-Delivery/blob/a24efce4260285bbf535f728dc6851e0b8194f96/evaluations/supersig_1_takahser.md?plain=1#L72), so feel free to add them.
Ok that has been added in `Tests`. [see pallet_supersif README](https://github.com/kabocha-network/pallet_supersig/blob/polkadot-v0.9.28/README.md)

Regarding the docker network issue, i will need to look into ubuntu and i am assuming you are port forwarding from your ubuntu server to your local machine's browser? If so what is your local machine?

If possible, and if its not too inconvenient for you, while i try and fix that issue: can you `cargo build --release` and then check the actual pallet? And by the time you have completed (or before) i shall have a solution for the docker.

**Comment by decentration** [comment]
@takahser re: Networking issue, if you check your browser, which probably doesn’t allow insecure ws (web socket) connections. You need to allow your browser to accept insecure connections, else if you have a nginx proxy to run it through wss:// then it will definitely work. 

If you check the second answer of this issue you can see how to allow `insecure content`: 

https://substrate.stackexchange.com/questions/4931/docker-container-p-99449944-running-node-template-not-working-on-polkadot-js/5014#5014

**Comment by decentration** [comment]
Though it needs to be said, the docker networking issues are nothing really to do with Substrate or pallet_supersig. If you try a paritytech docker image you will get the same issue. I thought it might be a docker command I was missing, but even consulting with Chevdor, who is Parity’s “docker whisperer” he validated that I have put all the correct tags to the command.

**Comment by takahser** [comment]
@decentration

> i am assuming you are port forwarding from your ubuntu server to your local machine's browser? If so what is your local machine?

That's correct, locally I'm using a MacBook Pro, running on an M1 chip.

> can you cargo build --release and then check the actual pallet supersig?

What exactly do you mean by "check the actual pallet supersig"?

> If you check the [second answer](https://substrate.stackexchange.com/questions/4931/docker-container-p-99449944-running-node-template-not-working-on-polkadot-js/5014#5014) of this issue you can see how to allow insecure content

I tried that, but it doesn't make a difference. It still doesn't connect.

image

**Comment by decentration** [comment]
@takahser it’s a docker networking problem, to guruantee it working through docker you will need a nginx reverse proxy so it runs with a secure connection (wss://). However, I think it’s overkill for testing. 

Can you agree it’s a Docker problem and outside the scope of this milestone? If you run a parity tech image you will get the same issue. 

Can you just compile without Docker so that you can get it to run on Polkadot js apps? the focal point of this milestone is pallet supersig. 

Hey Bill @laboon could you confirm this approach?

**Comment by decentration** [comment]
@takahser could you compile the branch so you can test all the other points? because we don’t need to let this Docker issue block the rest of the testing. We’re approaching week 7 so I’m hoping this is sorted soon.

**Comment by takahser** [comment]
@decentration I usually don't have this issue with other deliveries, but I understand your argument and you might be right on that. I'll compile it locally without docker and try to connect to it on polkadotJs apps, as you suggested. I'll get back to you soon.

**Review by takahser — APPROVED** [review]
@decentration thanks for your patience, I was able to run the node and I successfully smoke-tested the functionality using polkadotJs apps. I also took a look at your code, it looks clean and I'm happy about the extensive explanatory comments you added.
I'm approving this delivery, you can find my evaluation here: https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/supersig_1_takahser.md

**Comment by decentration** [comment]
@takahser thanks for reviewing it, am glad there was no further issue apart from the darned Docker. All the best : )

**Pull request merged**